src/frontend/src/optimizer/plan_node/stream_asof_join.rs

```diff
@@ -0,0 +1,387 @@
+            if left_input_ref.index() < left_input_len && right_input_ref.index() >= left_input_len
+            {
+                Ok(AsOfJoinDesc {
+                    left_idx: left_input_ref.index() as u32,
+                    right_idx: (right_input_ref.index() - left_input_len) as u32,
+                    inequality_type: Self::expr_type_to_comparison_type(expr_type)?.into(),
+                })
+            } else {
+                bail!("inequal condition from the same side should be push down in optimizer");
+            }
+        } else {
+            Err(ErrorCode::InvalidInputSyntax(
+                "AsOf join requires exactly 1 ineuquality condition".to_string(),
+            )
+            .into())
+        }
+    }
+
+    fn expr_type_to_comparison_type(expr_type: PbType) -> Result<PbAsOfJoinInequalityType> {
+        match expr_type {
+            PbType::LessThan => Ok(PbAsOfJoinInequalityType::AsOfInequalityTypeLt),
+            PbType::LessThanOrEqual => Ok(PbAsOfJoinInequalityType::AsOfInequalityTypeLe),
+            PbType::GreaterThan => Ok(PbAsOfJoinInequalityType::AsOfInequalityTypeGt),
+            PbType::GreaterThanOrEqual => Ok(PbAsOfJoinInequalityType::AsOfInequalityTypeGe),
+            _ => Err(ErrorCode::InvalidInputSyntax(format!(
+                "Invalid comparison type: {}",
+                expr_type.as_str_name()
+            ))
+            .into()),
+        }
+    }
+
+    /// Get join type
+    pub fn join_type(&self) -> JoinType {
+        self.core.join_type
+    }
+
+    /// Get a reference to the batch hash join's eq join predicate.
+    pub fn eq_join_predicate(&self) -> &EqJoinPredicate {
+        &self.eq_join_predicate
+    }
+
+    pub(super) fn derive_dist(
+        left: &Distribution,
+        right: &Distribution,
+        logical: &generic::Join<PlanRef>,
+    ) -> Distribution {
+        match (left, right) {
+            (Distribution::Single, Distribution::Single) => Distribution::Single,
+            (Distribution::HashShard(_), Distribution::HashShard(_)) => match logical.join_type {
+                JoinType::Unspecified
+                | JoinType::FullOuter
+                | JoinType::Inner
+                | JoinType::LeftOuter
+                | JoinType::LeftSemi
+                | JoinType::LeftAnti
+                | JoinType::RightSemi
+                | JoinType::RightAnti
+                | JoinType::RightOuter => unreachable!(),
+                JoinType::AsofInner | JoinType::AsofLeftOuter => {
+                    let l2o = logical
+                        .l2i_col_mapping()
+                        .composite(&logical.i2o_col_mapping());
+                    l2o.rewrite_provided_distribution(left)
+                }
+            },
+            (_, _) => unreachable!(
+                "suspicious distribution: left: {:?}, right: {:?}",
+                left, right
+            ),
+        }
+    }
```

Comment:
Can we merge this with `StreamHashJoin`?

Reply:
No. JoinType other than asof are unreachable here

Reply:
I mean, the method with the same name in `StreamHashJoin` basically do the same thing on the other match arms.

Reply:
You mean we should add a field `asof` and check in every JoinType branch see if it's acceptable? There is no strong reason to do this.

Reply:
Sure I can do this.